File: apps/boxc/src/pump/into/fun.cpp
#include "../../type.hpp"
#include <cstring>
#include <string_view>

Address get_into(Options options) {
    if (!options) return nullptr;
    std::string_view opt(options);
    auto pos = opt.find("into:");
    if (pos != std::string_view::npos) {
        return options + pos + 5;
    }
    if (opt.rfind("{", 0) == 0 || opt.rfind("[", 0) == 0) {
        return nullptr;
    }
    if (!opt.empty() && opt != "once" && opt != "many" && opt != "none") {
        return options;
    }
    return nullptr;
}

bool has_verb(Options options, const char* verb) {
    if (!options) return std::string_view(verb) == "GET";
    std::string_view opt(options);
    auto pos = opt.find("verb:");
    if (pos != std::string_view::npos) {
        auto rest = opt.substr(pos + 5);
        auto comma = rest.find(',');
        if (comma != std::string_view::npos) rest = rest.substr(0, comma);
        return rest == verb;
    }
    return std::string_view(verb) == "GET";
}

File: apps/rata/src/pump/into/fun.rs
use crate::r#type::{Address, Options};
use core::ffi::CStr;

pub fn get_into(options: Options) -> Option<Address> {
    if options.is_null() {
        return None;
    }
    let opt = unsafe { CStr::from_ptr(options).to_bytes() };
    if let Some(pos) = opt.windows(5).position(|w| w == b"into:") {
        return Some(unsafe { options.add(pos + 5) });
    }
    if opt.starts_with(b"{") || opt.starts_with(b"[") {
        return None;
    }
    if !opt.is_empty() && opt != b"once" && opt != b"many" && opt != b"none" {
        return Some(options);
    }
    None
}

pub fn has_verb(options: Options, verb: &[u8]) -> bool {
    if options.is_null() {
        return verb == b"GET";
    }
    let opt = unsafe { CStr::from_ptr(options).to_bytes() };
    if let Some(pos) = opt.windows(5).position(|w| w == b"verb:") {
        let rest = &opt[pos + 5..];
        let vlen = rest.iter().position(|&b| b == b',' || b == 0).unwrap_or(rest.len());
        return &rest[..vlen] == verb;
    }
    verb == b"GET"
}
